cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -153,8 +157,68 @@ protected void startUp() throws Exception {
     }
     delegate = new CompositeService(children);
-
     delegate.startAndWait();
+    // Explicitly emit both launching and running counts on startup.
+    emitFlowControlMetrics();
+  }
+
+  private void emitFlowControlMetrics(){
+    runRecordMonitorService.emitLaunchingMetrics();
+    runRecordMonitorService.emitRunningMetrics();
+  }
+
+  private void restoreActiveRuns(){
+    LOG.info("Restoring active runs");
+    int batchSize = cConf.getInt(Constants.RuntimeMonitor.INIT_BATCH_SIZE);
+    RetryStrategy retryStrategy =
+        RetryStrategies.fromConfiguration(cConf, Constants.Service.RUNTIME_MONITOR_RETRY_PREFIX);
+    long startTs = System.currentTimeMillis();
+
+    AtomicBoolean launching = new AtomicBoolean(false);
+    Retries.runWithRetries(
+        () ->
+            store.scanActiveRuns(
+                batchSize,
+                (runRecordDetail) -> {
+                  if (runRecordDetail.getStartTs() > startTs) {
+                    return;
+                  }
+                  try {
+                    LOG.info("Found active run: {}",runRecordDetail.getProgramRunId());
+                    if (runRecordDetail.getStatus() == ProgramRunStatus.PENDING) {
+                      launching.set(true);
+                      runRecordMonitorService.addRequest(runRecordDetail.getProgramRunId());
+                    } else if (runRecordDetail.getStatus() == ProgramRunStatus.STARTING) {
+                      launching.set(true);
+                      runRecordMonitorService.addRequest(runRecordDetail.getProgramRunId());
+                      // It is unknown what is the state of program runs in STARTING state.
+                      // A STARTING message is published again to retry STARTING logic.
+                      ProgramOptions programOptions =
+                          new SimpleProgramOptions(
+                              runRecordDetail.getProgramRunId().getParent(),
+                              new BasicArguments(runRecordDetail.getSystemArgs()),
+                              new BasicArguments(runRecordDetail.getUserArgs()));
+                      LOG.debug("Retrying to start run {}.", runRecordDetail.getProgramRunId());
+                      programStateWriter.start(
+                          runRecordDetail.getProgramRunId(),
+                          programOptions,
+                          null,
+                          this.store.loadProgram(runRecordDetail.getProgramRunId().getParent()));
+                    }
+                  } catch (Exception e) {
+                    ProgramRunId programRunId = runRecordDetail.getProgramRunId();
+                    LOG.warn(
+                        "Retrying to start run {} failed. Marking it as failed.", programRunId, e);
+                    programStateWriter.error(programRunId, e);
+                  }
+                }),
+        retryStrategy,
+        e -> true);
+    if (!launching.get()) {
```

Comment:
You don't need this anymore as you emit it separately anyway

Reply:
+1.

Reply:
Removed